### DISCUSSION TITLE
Pyinstaller packaged on python 3.13 shows DeprecationError

### DISCUSSION THREAD

**Discussion by baseplate-admin** [pyinstaller/pyinstaller#8920] [Help] [answered]
Hi,

I got this DeprecationWarning when packaging [an application](https://github.com/baseplate-admin/Shiinobi/) using the following `cli.spec`:

```python
 -*- mode: python ; coding: utf-8 -*-
a = Analysis(
    ['./src/shiinobi/cli.py'],
    pathex=['./src/shiinobi'],
    binaries=[],
    datas=[],
    hiddenimports=[],
    hookspath=['./src/_pyinstaller'],
    hooksconfig={},
    runtime_hooks=[],
    excludes=[],
    noarchive=True,
    optimize=2,
)
pyz = PYZ(a.pure)

exe = EXE(
    pyz,
    a.scripts,
    a.binaries,
    a.datas,
    [],
    name='shiinobi',
    debug=False,
    bootloader_ignore_signals=False,
    strip=False,
    upx=True,
    upx_exclude=[],
    runtime_tmpdir=None,
    console=True,
    icon='./assets/shiinobi.png',
    disable_windowed_traceback=False,
    argv_emulation=False,
    target_arch=None,
    codesign_identity=None,
    entitlements_file=None,
)
```

```
PyInstaller\hooks\rthooks\pyi_rth_pkgres.py:37: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
```

I was wondering if this can be addressed or can i help in any way?

Thank you

**Comment by baseplate-admin** [comment]
If anyone wants an executable,

https://github.com/baseplate-admin/Shiinobi/actions/runs/12329900575

**Comment by bwoodsend** [comment, marked as answer]
That deprecation warning will be issued if your application contains `pkg_resources` so if you want rid of the warning, then you need to get rid of `pkg_resources`. It might be that `pkg_resources` is getting picked up unnecessarily – you can try `--exclude-module=pkg_resources` and either the warning will go away or you'll discover which part of your dependency tree is using deprecated functionality.

**Reply by baseplate-admin** [reply]
Hi, thank you for quick response, 

I think the issue is From here (not from my package using `pkg_resource` because i dont think i use `pkg_resource` anywhere)

https://github.com/pyinstaller/pyinstaller/blob/57ae1ed9cd51b1be55d14ae25da86361afdbf28f/PyInstaller/hooks/rthooks/pyi_rth_pkgutil.py#L17

---
> if you want rid of the warning, then you need to get rid of pkg_resources.

When building without `pkg_resource` everything still works as expected. Thank you for your insight in this.

**Reply by bwoodsend** [reply]
If you can retrieve the `build/xref-*.html` file that PyInstaller produces, that can tell you what other module it thought needs `pkgutil` and/or `pkg_resources`. It might be that this false positive can be suppressed generally for that module.

**Reply by baseplate-admin** [reply]
> If you can retrieve the build/xref-*.html file that PyInstaller produces, that can tell you what other module it thought needs pkgutil and/or pkg_resources. It might be that this false positive can be suppressed generally for that module.

Yep i see the issue, Thank you for helping me debug this :) Cheers!

**Reply by baseplate-admin** [reply]
[xref-cli.html.odt](https://github.com/user-attachments/files/18141386/xref-cli.html.odt)

Here is the xref-cli.html (remove the odt) it seems pyinstaller picked up `packaging` and `pkg_resources` along with it, i dont know why it might've been the case

**Reply by rokm** [reply]
Hmm, looks like `pkg_resources` is mainly referenced from `setuptools` modules in this case. There is a `pkg_resources` import in `setuptools.dist` (which in turn is imported from top-level `setuptools`) that happens under `if TYPE_CHECKING`. There are some more imports that *seem* to be direct (e.g., in `setuptools.installer`), but these modules, in turn, are delay-imported in specific codepaths. So I suppose we might get away with excluding imports of `pkg_resources` in top-level `setuptools` hook (perhaps to be safe, only from version that added deprecation warning).

There's another reference seemingly coming from `packaging`, but that one is our fault - https://github.com/pyinstaller/pyinstaller/blob/develop/PyInstaller/hooks/hook-packaging.py . I suspect this was added to potentially accommodate `pkg_resources`-vendored versions of `packaging` (`pkg_resources._vendor.packaging`). But I don't think that vendored package was ever exported as global top-level package, so the hook is (and has been) unnecessary. So we should probably drop it.

**Comment by gtoombs-avidrone** [comment]
On a "current" environment with Python 3.12.3 and

```none
pyinstaller==6.13.0
pyinstaller-hooks-contrib==2025.4
pytz==2025.2
```

the xref.html says:

> `pkg_resources` Package
>
> imported by: `packaging` `pyi_rth_pkgres.py` `pytz`

Since I am writing a .spec manually, I add `excludes=['pkg_resources']` to `Analysis()` and the message goes away.

**Reply by rokm** [reply]
> imported by: packaging pyi_rth_pkgres.py pytz

In this case, it is pulled in because it is imported somewhere in `pytz`.

**Reply by gtoombs-avidrone** [reply]
Sure, though (perhaps by luck) I do not exercise any code path that requires `pkg_resources`, so exclusion is still the best route for me.

**Reply by rokm** [reply]
Indeed; it looks like it is used in a fallback codepath that is normally not triggered:

https://github.com/stub42/pytz/blob/82e0891730a38fdcf8c9c680af34712d45a97fde/src/pytz/__init__.py#L78-L108

We should probably add an exclude to out `pytz` hook to avoid collecting it via this particular import chain.
